Introduce CloudEvent to KEDA (#1227)
kedacore/keda-docs: content/docs/2.13/operate/events.md

| `KEDAScaleTargetDeactivationFailed`   | `Warning` | When KEDA fails to scale the scale target of a ScaledObject to 0| NO |                                                              |
| `KEDAJobsCreated`                     | `Normal`  | When KEDA creates jobs for a ScaledJob | NO |                                                                                       |
| `TriggerAuthenticationAdded`          | `Normal`  | When a new TriggerAuthentication is added| NO |                                                                                     |
| `TriggerAuthenticationDeleted`        | `Normal`  | When a TriggerAuthentication is deleted| NO |                                                                                       |
| `ClusterTriggerAuthenticationAdded`   | `Normal`  | When a new ClusterTriggerAuthentication is added| NO |                                                                              |
| `ClusterTriggerAuthenticationDeleted` | `Normal`  | When a ClusterTriggerAuthentication is deleted| NO |                                                                                |

## CloudEvent Support (Experimental)

> ### CloudEventSource Resource

Reviewer: ```suggestion
### Subscribing to events with `CloudEventSource`
```

> `CloudEventSource` resource now can be created in KEDA for emitting events to user's custom CloudEvent sink. Event will be emitted to both Kubernetes Events and CloudEvents Destination if CloudEventSource resource is created. Here is a the schema of the `CloudEventSource` CRD:

Reviewer: ```suggestion
`CloudEventSource` resource can be used in KEDA for subscribing to events that are emitted to the user's defined CloudEvent sink.

> 📝 Event will be emitted to both Kubernetes Events and CloudEvents Destination if CloudEventSource resource is created.

Here is a the schema of the `CloudEventSource` CRD:
```

```yaml

> apiVersion: eventing.keda.k8s.io/v1alpha1

Reviewer 2: The apiVersion is not correct

kind: CloudEventSource
metadata:
  name: {cloud-event-name}
spec:
  clusterName: {cluster-name} #Optional. Will be used in the source/subject to specify where the event comes from. The default value is 'kubernetes-default' and it can also be set during the installation of KEDA with --k8sClusterName. This one will overwrite others if set.
  destination:
    http:
      uri: http://foo.bar
```

In general, an event emitted by KEDA would fundamentally come down to the following structure:
```json
{
    "specversion" : "1.0",
    "type" : "com.cloudeventsource.keda",
    "source" : "/{cluster-name}/{namespace}/keda",
    "subject" : "/{cluster-name}/{namespace}/workload/{scaledobject-name}",
    "id" : "<guid>",
    "time" : "2018-04-05T17:31:00Z",
    "datacontenttype" : "application/json",
    "data" : {
      "reason":"<event-reason>",
      "message":"<event-message>"
   }
}
```

> ### Destination
> There will be multiple type of destination to emit KEDA events. Nowadays an HTTP CloudEvent destination is supported.
> #### CloudEvent HTTP

Reviewer: Massive nit
```suggestion
### Event Sinks

There will be multiple types of destination to emit KEDA events to.

Here is an overview of the supported destinations:

- [HTTP endpoint](#http-endpoint).

#### HTTP endpoint
```

| `TriggerAuthenticationDeleted`        | `Normal`  | When a TriggerAuthentication is deleted                                                                                     |
| `ClusterTriggerAuthenticationAdded`   | `Normal`  | When a new ClusterTriggerAuthentication is added                                                                            |
| `ClusterTriggerAuthenticationDeleted` | `Normal`  | When a ClusterTriggerAuthentication is deleted                                                                              |

> ## CloudEvent Support (Experimental)

Reviewer: Can we document the supported events end-users can subscribe to please?

### CloudEventSource Resource

> `CloudEventSource` resource now can be created in KEDA for emitting events to user's custom CloudEvent sink. Event will be emitted to both Kubernetes Events and CloudEvents Destination if CloudEventSource resource is created. This specification describes the `CloudEventSource` Custom Resource definition:

Reviewer: > Event will be emitted to both Kubernetes Events and CloudEvents Destination if CloudEventSource resource is created.

I personally think we should not rely on this CRD to decide if we emit these events to Kubernetes Events, can we please change the implementation for this?

> ## CloudEvent Support (Experimental)
> 
> ### CloudEventSource Resource
> `CloudEventSource` resource now can be created in KEDA for emitting events to user's custom CloudEvent sink. Event will be emitted to both Kubernetes Events and CloudEvents Destination if CloudEventSource resource is created. This specification describes the `CloudEventSource` Custom Resource definition:

Reviewer: Let's polish this a bit
```suggestion
## CloudEvent Support (Experimental)

### Subscribing to CloudEvents with our CloudEventSource CRD

`CloudEventSource` allows you to subscribe to one or more events in KEDA and let it emit them to user's event sink and use payloads that are compliant with the [CloudEvent specification](https://cloudevents.io/).
```

> [`cloudeventsource_types.go`](https://github.com/kedacore/keda/blob/v2.13.0/pkg/apis/keda/v1alpha1/cloudeventsource_types.go)

Reviewer: I don't think we should link to code here
```suggestion
Here is a the schema of the `CloudEventSource` CRD:
```

```yaml
apiVersion: eventing.keda.k8s.io/v1alpha1
kind: CloudEventSource
metadata:
  name: {cloud-event-name}
spec:
  clusterName: {cluster-name} #Optional. Will be used in source/subject. And a "default" value will be set if this value is empty 
  destination:
    http:
      uri: http://foo.bar
```

In general, an event emitted by KEDA would fundamentally come down to the following structure:
```json
{
    "specversion" : "1.0",
    "type" : "com.cloudeventsource.keda",
    "source" : "/{cluster-name}/{namespace}/keda",
    "subject" : "/{cluster-name}/{namespace}/workload/{scaledobject-name}",
    "id" : "<guid>",
    "time" : "2018-04-05T17:31:00Z",
    "datacontenttype" : "application/json",
    "data" : {
      "reason":"<event-reason>",
      "message":"<event-message>"
   }
}
```

> ### Destination
> There will be multiple type of destination to emit KEDA events. Nowadays an HTTP CloudEvent destination is supported.
> #### CloudEvent HTTP

Reviewer: ```suggestion
### Event Destinations

Here is an overview of the event destinations that KEDA supports:

- [HTTP endpoint](#http-endpoint)

#### HTTP endpoint
```
